client/src/main/scala/com/redi/scalademo/presentation/FormListener.scala

```diff
@@ -0,0 +1,105 @@
+package com.redi.scalademo.presentation
+
+import autowire.{clientCallable, unwrapClientProxy}
+import com.redi.scalademo.business.{Calculator, NumericStringValidator, ValidatedResult}
+import com.redi.scalademo.infrastructure.Client
+import org.scalajs.dom
+import org.scalajs.jquery.{JQuery, JQueryEventObject, jQuery ⇒ $}
+
+import scala.concurrent.Future
+import scala.math.BigDecimal
+import scala.scalajs.concurrent.JSExecutionContext.Implicits.queue
+import scala.scalajs.js
+
+class FormListener(
+  client: Client,
+  numericStringValidator: NumericStringValidator
+) {
+
+  def attachTo(formId: String): Unit = {
+
+    val InfoClass = "is-info"
+    val DangerClass = "is-danger"
+
+    val Validate = "validate"
+    val Augend = "augend"
+    val Addend = "addend"
+    val Summand = "summand"
+
+    val Name = "name"
+    val Value = "value"
+
+    val formElement: JQuery = $(s"form#$formId")
+    val summandElement: JQuery = formElement.find(elementNamed(Summand))
+
+    formElement.submit { (e: JQueryEventObject) ⇒
+
+      e.preventDefault()
+
+      formElement.find(namedElement).removeClass(DangerClass)
+      val serializedFormData = formElement.serializeArray().asInstanceOf[js.Array[js.Dictionary[String]]]
+      val (validationFormData: js.Array[js.Dictionary[String]], calculableFormData: js.Array[js.Dictionary[String]]) =
+        serializedFormData.partition(_(Name) == Validate)
+
+      val shouldValidate: Boolean = validationFormData.forall(_(Value).nonEmpty)
+
+      var allFormValuesAreValid = true
+      var augend: String = ""
+      var addend: String = ""
+
+      for (formControlData: js.Dictionary[String] ← calculableFormData) {
+        val name: String = formControlData(Name)
+        val value: String = formControlData(Value)
+        if (shouldValidate) {
+          val isValid: Boolean = numericStringValidator.isValidNumber(value)
+          if (!isValid) {
+            allFormValuesAreValid = false
+            formElement.find(elementNamed(name)).addClass(DangerClass)
+          }
+        }
+        name match {
+          case Augend ⇒
+            augend = value
+          case Addend ⇒
+            addend = value
+        }
+      }
+
+      if (allFormValuesAreValid) {
+        val futureResult: Future[ValidatedResult[BigDecimal]] =
+          client[Calculator].add(augend, addend).call()
+        for (result: ValidatedResult[BigDecimal] ← futureResult) {
+          result match {
+            case Left(validationFailures) ⇒
+              for (validationFailure ← validationFailures) {
+                formElement.find(elementNamed(validationFailure.formControlName)).addClass(DangerClass)
+              }
+            case Right(summand: BigDecimal) ⇒
+              summandElement
+                .value(summand.toString)
+                .addClass(InfoClass)
+          }
+        }
+      }
+
+      ()
```

Comment:
I can leave it out if you think it's too distracting. The callback expects a return type of `js.Any`, and the expression above this one, being a `for` loop, returns `Unit`, so an implicit conversion was happening between the two, creating a big gray underline in IDEA. I thought to avoid the implicit conversion by explicit returning a Unit, which is what this `()` is.